migrate to fabric-keybindings-v1 and remove builder

shedaniel opened the pull request:
This is an edited clone of #244 as that seemed dead and I want to push stuff going faster.

- Fixed checkstyle issues
- Migrated to fabric-keybindings-v1
- Added a test mod
- Added FabricKeyBinding#getBoundKeyOf for getting vanilla bound keys with ease
- Renamed `registered` to `automaticallyRegister` as that is more of a better name
- Added a couple of Objects.requireNonNull validations

shedaniel committed:
Edited Clone of #244
- Fixed checkstyle issues
- Migrated to fabric-keybindings-v1
- Removed sticky keybindings from #244 as it sounds useless and you can just around it by simply adding that functionality yourself, I might add it back if someone can tell me the advantages of sticky keys except bloat
- Added a test mod
- Added FabricKeyBinding#getBoundKeyOf for getting vanilla bound keys with ease
- Renamed `registered` to `automaticallyRegister` as that is more of a better name
- Added a couple Objects.requireNonNull validations

shedaniel committed:
Add back StickyFabricKeyBinding as it is in vanilla, did not notice.

shedaniel committed:
Remove extra "key."

i509VCB commented:
I know this is a client only api, but it would be nice if there was a `getKeybindText` method that returns the key bind text that corresponds to keybinding

i509VCB commented on fabric-keybindings-v1/build.gradle:
```diff
@@ -0,0 +1,7 @@
+archivesBaseName = "fabric-keybindings-v1"
+version = getSubprojectVersion(project, "0.1.0")
```
1.0.0

i509VCB commented on fabric-keybindings-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,219 @@
+		 *
+		 * <p>Mods who intend to register their own key bindings manually may choose not to use this.</p>
+		 */
+		public Builder automaticallyRegister() {
```
Hmm, maybe instead of this, we have a `build` and a `buildAndRegister` method

shedaniel replied:
I would do build().register() if that's nice

shedaniel committed:
Bump to 1.0.0

shedaniel committed:
build().register()

liach commented on fabric-keybindings-v0/src/main/java/net/fabricmc/fabric/api/client/keybinding/FabricKeyBinding.java:
```diff
@@ -20,27 +20,31 @@
+	@Override
 	public InputUtil.KeyCode getBoundKey() {
-		return ((KeyCodeAccessor) this).getKeyCode();
+		return super.getBoundKey();
```
Can probably remove this override?

shedaniel replied:
will it preserve backwards compatibility if I do this? (I am not sure so I did this)

i509VCB replied:
Make the v0 use a delegate class internally like how I did FabricBlockSettings v1

shedaniel replied:
but the v0 one extends Keybinding anyways

liach replied:
It should preserve backward compatibility... the call to old keybinding's `getBoundKey` should be invokevirtual in existing bytecode

Hmm, I guess we can give the new method a new name like `getDefaultCode` instead

shedaniel replied:
is not the default code, it returns the current code, bound key is the best name for this imo

liach commented on fabric-keybindings-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,214 @@
+	 *
+	 * <p>May be different from the {@link #getId()}.</p>
+	 */
+	public Identifier getIdentifier() {
```
```suggestion
	public Identifier getId() {
```

Juuxel replied:
> May be different from the {@link #getId()}.

There already seems to be one.

liach replied:
Hmm, then need to fix that in yarn probably. In api that thing is called `translationKey`, which should be right if it starts with `key.`
https://maven.fabricmc.net/docs/yarn-20w19a+build.1/net/minecraft/client/options/KeyBinding.html#KeyBinding-java.lang.String-net.minecraft.client.util.InputUtil.Type-int-java.lang.String-

shedaniel replied:
I'll keep this because this PR is for 1.15

liach commented on fabric-keybindings-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,214 @@
+	 */
+	public <T extends FabricKeyBinding> T register() {
+		KeyBindingRegistry.INSTANCE.register(this);
+		return (T) this;
```
I personally disagree with the addition of this method. Imo error-prone and violating oop design.

liach replied:
@shedaniel the other two points are rejected but this you need to explain or change. Imo this is dangerous as it allows random error-prone casts, just ask users call `KeybindingRegistry.INSTANCE.register` on their own

shedaniel replied:
will do.

shedaniel committed:
Remove `register()`

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Fix test

Signed-off-by: shedaniel <daniel@shedaniel.me>

liach reviewed (approved):
Should be good, guess your pr may wait for the mapping update and have a proper `getId` method

shedaniel commented:
Might want to create a 1.16 PR of this or also make a mapping update for 1.15

Prospector commented:
Rename to `fabric-key-binding-api-v` to fit the naming scheme

shedaniel commented:
I would choose `fabric-keybinding-api-v1` because key.binding package sounds weird.

Prospector commented:
Package would remain `keybinding` because package names can consist of multiple words, while module names must have every word separated.

Refer to the specification which clearly lays this out and was unanimously agreed upon: #575

i509VCB commented:
A GameRule API would be `fabric-gamerule-api-v1` since a gamerule is an actual concept. We wouldn't do `fabric-game-rule-api-v1`

Is a `keybind` an actual concrete concept or is a subset of `key` where it would be `key-bind`

Prospector commented:
Module names do not care whether it is a concept or not. Every word is separated by a hypen. The concrete concept part only applies to package names (all this is written out in #575).

The important parts:
> Every name consisting of multiple words will have a hypen (`-`) between the words. This includes class names such as `BlockEntity` being represented as `block-entity`.

>- A module named `fabric-networking-block-entity-v1` would have the root package `net.fabricmc.fabric.api.networking.blockentity`, because "Block Entity" is a singular construct under "networking", and should not imply that `entity` was hierarchically inferior to `block`

shedaniel commented:
Fair enough, will do

shedaniel committed:
Rename module

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Fix checkstyle violation

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel commented:
we might as well make a 2021 license pr and we might be able to merge it in time next year /s

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+ * FabricKeyBinding right = builder
+ * 			.id(new Identifier("example", "right"))
+ * 			.code(Keys.Right)
+ * 			.build();
```
Is there a specific reason to reuse the builder instance? It may be nicer to supply the required parameter (id) on creation, it is only one after all and recreating the builder doesn't look too bad to me.

shedaniel replied:
you shouldn't reuse the builder instance, will update this example

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+ * KeyBindingRegistry.register(right);
+ * </code></pre>
+ */
+public class FabricKeyBinding extends KeyBinding {
```
final?

shedaniel replied:
need to extend this for v0.

Player3324 replied:
Any reasonable way to avoid this? Allowing sub classes may make future changes more difficult.

shedaniel replied:
Would making an interface for detecting if it is a fabric keybind be alright? need to do that for it to work rn, might look at how it works to tweak it

Player3324 replied:
If there's no simple solution we can leave it not final

shedaniel replied:
Made an interface called `ModdedKeyBinding` for both the v1 and v0 keybinding to extend, I can't think of a better name for that class.

shedaniel replied:
nvm I reverted this.

shedaniel replied:
We might be above to just never return the fabric version of the keybind to the player, and return the vanilla ones, I don't see a clear reason how getIdentitier() is useful, vanilla keybindings don't have them anyways

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+	 */
+	public static Builder builder() {
+		return new Builder();
+	}
```
This method should be first or last, depending on what we do elsewhere

shedaniel replied:
Should be first, I changed the javadoc to make it not reuse the builder and you shouldn't be.

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+ * 			.code(Keys.Right)
+ * 			.build();
+ * KeyBindingRegistry.register(left);
+ * KeyBindingRegistry.register(right);
```
Wouldn't a .buildAndRegister be better than exposing and requiring the registration separately? Is there any good reason not to want to register directly?

shedaniel replied:
I am making this register both the keybinding and the category automatically now

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+		}
+
+		/**
+		 * Indicates to this builder that keybindings built through it are intended to be unbound.
```
Maybe something like "... aren't assigned to any key by default" to more clearly state the effect?

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+		 * @param keyCode The default key code. Must be a valid key. May not be -1.
+		 */
+		public Builder code(int keyCode) {
+			Preconditions.checkState(keyCode != UNASSIGNED, "UNASSIGNED is not a valid key code.");
```
Is the any way we can verify the integer more? Maybe check against some reasonable lower and upper bound?

shedaniel replied:
Just checked, it seems like there isn't a limit on what the integer can be, Minecraft will just create a new code if it doesn't exist already, meaning that it can basically accept any code.

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+		}
+
+		/**
+		 * Sets the key's type. Maybe be one of [{@link InputUtil.Type#KEYSYM} (keyboard), {@link InputUtil.Type#SCANCODE}, {@link InputUtil.Type#MOUSE}]
```
Does MC not allow switching key binds between mouse and keyboard?

shedaniel replied:
this is the type of the code specified above, this is just for the default binding

Player3324 replied:
Can we combine this with the code (int) method then, maybe having per-type methods?

shedaniel replied:
I'll just combine it, don't know if a mod would like to extend the enum

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/KeyBindingRegistry.java:
```diff
@@ -0,0 +1,44 @@
+	 * @param categoryName The key binding category name.
+	 * @return True if a new category was added.
+	 */
+	boolean addCategory(String categoryName);
```
If we add register to the builder's build(), this method could get moved as a static helper to KeyCategories.

shedaniel replied:
should we just automatically register the category if it doesn't exist?

shedaniel replied:
Seems like we are already doing that with a warning that doesn't make sense, I think I will just remove that warning.

dexman545 replied:
Just registering the category makes sense

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+	 *
+	 * @return configured KeyCode
+	 */
+	public InputUtil.KeyCode getBoundKey() {
```
There are some naming inconsistencies between bound vs assigned.

shedaniel replied:
We have been using bound in the new yarn, where is assigned used

Player3324 replied:
Search the file for "assign", there are multiple uses both in code and docs

Player3324 commented on fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java:
```diff
@@ -0,0 +1,204 @@
+		 *
+		 * @param toggleFlagSupplier A getter function to determine whether to toggle or not. True for toggling behaviour, false otherwise.
+		 */
+		public Builder sticky(BooleanSupplier toggleFlagSupplier) {
```
I'd use a little nested class instead of BooleanSupplier that can then use a better method name and document it.

Player3324 reviewed (commented):
Looks good overall, I found a few things to tweak or consider. Additionally, is there anything worth doing about conflicting key binds?

shedaniel committed:
major refactor

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
revert some stuff

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
fix build

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
major stuff

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel commented:
@sfPlayer1 I've completely removed `FabricKeyBinding` in v1 and returns `KeyBinding` for the builder.
The `builder()` method is now in the `KeyBindingUtil` class but I assume we can we better names without confusing with the original `FabricKeyBinding` name in v0.
BooleanSupplier got `StickyBindingProvider` and identifiers are not passed through the keybinding anymore because IMO they are useless now.
I've also removed a reflection in favour of Accessors, please review.

shedaniel committed:
fix license, of course

Signed-off-by: shedaniel <daniel@shedaniel.me>

i509VCB commented on fabric-key-binding-api-v1/build.gradle:
```diff
@@ -0,0 +1,7 @@
+
+dependencies {
+	compile project(path: ':fabric-api-base', configuration: 'dev')
+	testmodCompile project(path: ':fabric-events-lifecycle-v0', configuration: 'dev')
```
Do note because you have translation keys in your tests, you need to add the resource loader to the testmodCompile

shedaniel replied:
oh right.

shedaniel committed:
Add resource loader v0

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Let's not break the api.

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel committed:
Rename to buildAndRegister

Signed-off-by: shedaniel <daniel@shedaniel.me>

Player3324 reviewed (commented):
I think it's good now, just a couple trivial changes to consider:

- IMO nested class for StickyBindingProvider, it is trivial and only has one very specific use
- KeyBindingRegistryImpl may as well use static methods since it doesn't need to be an object (no iface)
- KeyBindingRegistryImpl.process could use a comment stating what it does and why
- We use Helper over Util

shedaniel committed:
resolve reviews

Signed-off-by: shedaniel <daniel@shedaniel.me>

Player3324 reviewed (approved):
Good now, Thanks!

modmuss50 commented on fabric-key-binding-api-v1/src/testmod/java/net/fabricmc/fabric/test/client/keybinding/KeybindingsTest.java:
```diff
@@ -0,0 +1,62 @@
+		KeyBinding binding1 = KeyBindingHelper.builder()
+				.identifier(new Identifier("fabric-key-binding-api-v1-testmod:test_keybinding_1"))
+				.category("category.first.test")
+				.key(InputUtil.Type.KEYSYM, 80) // P
```
Use the GLFW constants `GLFW.GLFW_KEY_P` will be clearer and act as a better example

shedaniel replied:
fixed.

shedaniel committed:
Use GLFW

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel commented:
Actually now when I think about it, why do we need a builder at all? We can just call the KeyBinding constructor and register that.

shedaniel commented:
What are the options if we just straight up removed the builder and provide the register method only.

liach commented:
I think the original problem with keybinding is that they are in a fixed-size array. If that is really the only thing restricting modded keybindings, we should just make it simple like object builder's default entity attribute registry.

shedaniel commented:
No, I meant to allow devs to use the vanilla constructor, while registering them with our custom thing to handle the array and category order.

The api we have right now is just a redirect to the vanilla constructor, and anyone can just do that.

shedaniel committed:
Dump the builder entirely

Signed-off-by: shedaniel <daniel@shedaniel.me>

shedaniel commented:
I've removed the builder, [example code](https://github.com/FabricMC/fabric/pull/615/files#diff-9a7d4026510ab1b66e9a5dc7ee0b4458R33-R35).

Also removed the dependency from `fabric-api-base`, turns out we don't use it except in the testmod.

modmuss50 commented:
This can be moved to last call if its ready.

Juuxel commented on fabric-key-binding-api-v1/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "schemaVersion": 1,
+  "id": "fabric-key-binding-api-v1",
+  "name": "Fabric Key Bindings API (v1)",
```
```suggestion
  "name": "Fabric Key Binding API (v1)",
```
Also, the name and ID use "key binding" but the description uses "keybinding", should be made consistent.

shedaniel replied:
Fixed

shedaniel committed:
Rename to Key Binding

Signed-off-by: shedaniel <daniel@shedaniel.me>
